src/js/components/DSLFormWithExpressionUpdates.js

```diff
@@ -0,0 +1,172 @@
+import React, {PropTypes} from 'react';
+
+import DSLCombinerTypes from '../constants/DSLCombinerTypes';
+import DSLFilterTypes from '../constants/DSLFilterTypes';
+import DSLUpdatePolicy from '../constants/DSLUpdatePolicy';
+import DSLExpression from '../structs/DSLExpression';
+import DSLUpdateUtil from '../utils/DSLUpdateUtil';
+import DSLUtil from '../utils/DSLUtil';
+import {FilterNode} from '../structs/DSLASTNodes';
+
+const METHODS_TO_BIND = [
+  'handleFormChange',
+  'handleFormSubmit'
+];
+
+/**
+ * This <form /> component wraps a set of <input /> elements and automates
+ * the update of the DSL expression based on the part definition given.
+ */
+class DSLFormWithExpressionUpdates extends React.Component {
+  constructor() {
+    super(...arguments);
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  /**
+   * Handle a change to the form
+   */
+  handleFormChange({target}) {
+    let {
+      expression,
+      groupCombiner,
+      itemCombiner,
+      onChange,
+      parts,
+      updatePolicy
+    } = this.props;
+    let value = target.value;
+
+    const name = target.getAttribute('name');
+    if (!name) {
+      return;
+    }
+
+    if (target.type === 'checkbox') {
+      value = target.checked;
+    }
+
+    // Locate the respective DSL part node
+    const partNode = parts[name];
+    if (partNode == null) {
+      return;
+    }
+
+    // Locate all the current tokens of the expression that the filter match
+    const matchingNodes = DSLUtil.findNodesByFilter(expression.ast, partNode);
+
+    switch (partNode.filterType) {
```

Comment:
Let's create a separate function for this and return an `expression`, we can pass to `onChange`